### PR TITLE
Automatically handle IBus configuration to avoid conflicts with KDE

### PR DESCRIPTION
**Pull request by ArrayBolt3**
Resolves the issue in internal ticket 4027. No user intervention should be needed beyond following a pop-up prompt that may appear on login for people with a misconfigured IBus installation.

### CODE REVIEW

File: package-main/usr/lib/kfocus/bin/kfocus-fix-ibus [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+# Example  : kfocus-fix-ibus check; kfocus-fix-ibus repair;
+# License  : GPLv2
+# Run By   : Developers only
+# Spec     : 4027
+#
+set -u;
+
+_fi2StderrFn () { 1>&2 echo -e  "$*"; }
+
+## BEGIN _fi2RepairIbusConfigFn {
```
Comment (mmikowski) [inline package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:22]: Fantastic documentation. If we create a rtest, we should move all this there. Add a comment to move to rtest when the time is right.

writing a rtest is actually quite simple for this; can mock qdbus...

File: package-main/usr/lib/kfocus/bin/kfocus-fix-ibus [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+}
+## . END _fi2CheckIbusConfigFn }
+
+## BEGIN _fi2MainFn {
+_fi2MainFn () {
+  declare _mode_str;
+
+  _mode_str="${1:-}";
+
+  if [ "${_mode_str}" = 'check' ]; then
```
Comment (mmikowski) [inline package-main/usr/lib/kfocus/bin/kfocus-fix-ibus]: 1. Provide a mode that checks and then applies if needed. 
2. For multi-branch checks, use case statements

```bash
case "${_mode_str}" in
  check) ;;
  repair) ;;
  *) ;; 
esac

File: package-main/usr/lib/kfocus/bin/kfocus-fix-ibus [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      return 1;
+    fi
+  fi
+}
+## . END _fi2CheckIbusConfigFn }
+
+## BEGIN _fi2MainFn {
+_fi2MainFn () {
+  declare _mode_str;
+
```
Comment (mmikowski) [inline package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:54]:
```bash
if [ -z "${_qdbusExe}" ]; then
  # Avoid pipelines whenever you can. This is almost certainly better like this:
  if grep -qa 'org.freedesktop.ibus' < <(qdbus); then
    _fi2StderrFn 'qdbus not found';
    return 1;
  fi
```
...

File: package-main/usr/lib/kfocus/bin/kfocus-fix-ibus [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+## . END _fi2MainFn }
+
+## BEGIN Run fi2MainFn when script is NOT sourced {
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
+  _binName="$(  readlink -f "$0"       )" || exit 101;
+  _binDir="$(   dirname  "${_binName}" )" || exit 101;
+  _baseDir="$(  dirname  "${_binDir}"  )" || exit 101;
+  _baseName="$( basename "${_binName}" )" || exit 101;
+
```
Comment (mmikowski) [inline package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:109]:
```
_qdbusExe="$(command -v qdbus)";
```

File: package-main/usr/lib/kfocus/bin/kfocus-fix-ibus [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      _fi2StderrFn 'IBus configuration repair failed!';
+      return 1;
+    fi
+  else
+    _fi2StderrFn "Unrecognized command |${_mode_str}|.";
+    return 1;
+  fi
+}
+## . END _fi2MainFn }
+
```
Comment (mmikowski) [inline package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:93]: Add declaration of package variables here
```bash
declare _binName  ...

**Review by mmikowski — CHANGES_REQUESTED** [review]
Looks excellent, but could you please make the few tweaks and then just let me know.  I'd appreciate it!

**Comment by ArrayBolt3** [comment]
I believe I have implemented the requested changes. I also adjusted the main worker script to make it easier to test, and got the regression tests in much better shape.

**Comment by ArrayBolt3** [comment]
I can't figure out how to get this marked as ready for review :-/ :P

**Pull request merged**